EQL: [Docs] Add documentation for the CircuitBreaker (#74897)
elastic/elasticsearch: docs/reference/eql/eql.asciidoc

You can also manually delete saved synchronous searches using the
<<delete-async-eql-search-api,delete async EQL search API>>.

[discrete]
[[eql-circuit-breaker]]
=== EQL circuit breaker settings

> When a <<eql-sequences, sequence>> query is executed, the node handling the query,

Reviewer: The last `,` looks misplaced.

> needs to keep some structures in memory which are needed by the algorithm which
> implements the sequence matching. When large amounts of data need to be processed,

Reviewer 2: > which are needed by the algorithm which implements

Too many `which`es?
How about
```suggestion
needs to keep some structures in memory that are needed by the algorithm implementing the sequence matching. When large amounts of data need to be processed,
```

and/or a large amount of matched sequences is requested by the user (by setting the
<<eql-search-api-params-size, size>> query param), the memory occupied by those

> structures could potentially exceed the available memory in the JVM. This would cause

Reviewer 2: ```suggestion
structures could potentially exceed the available memory of the JVM. This would cause
```

> When a <<eql-sequences, sequence>> query is executed, the node handling the query
> needs to keep some structures in memory, which are needed by the algorithm
> implementing the sequence matching. When large amounts of data need to be processed,
> and/or a large amount of matched sequences is requested by the user (by setting the
> <<eql-search-api-params-size, size>> query param), the memory occupied by those
> structures could potentially exceed the available memory of the JVM. This would cause
> an `OutOfMemory` exception which would bring down the node.
> 
> To prevent this from happening, a special <<circuit-breaker, circuit breaker>> is used,
> which limits the memory allocation during the execution of a <<eql-sequences, sequence>>
> query. When the breaker is triggered, an `org.elasticsearch.common.breaker.CircuitBreakingException`
> is thrown and a descriptive error message is returned to the user.

Reviewer 3: This copy is good, but I wonder if we're over-explaining a bit. I'd consider shortening it:

```suggestion
The EQL circuit breaker limits the amount of memory that sequence queries can
use. The breaker prevents an OutOfMemoryError for queries with a large
<<eql-search-api-params-size,`size`>> or data set.
```

Author: @costin What do you think?

Reviewer 2: I'd leave it as is. It didn't strike me as too verbose or too detailed. If the overall style of EQL documentation or that of the circuit breakers in Elasticsearch is more concise/trimmed down, then go with @jrodewig suggestion. Otherwise, my personal opinion is to leave it as is. Circuit breakers are a tricky subject in general imo since the memory calculation algorithm is rather  non-obvious to the regular user. Enough information passed on in the docs helps sometimes clearing out the confusion around them.

Reviewer: What @astefan wrote.

Author: Thx!

> query, so instead of an `OutOfMemory` exception, a `org.elasticsearch.common.breaker.CircuitBreakingException`
> is thrown, and a descriptive error message is returned to the user.

Reviewer 2: If the circuit breaker doesn't do its job, it doesn't mean an OOM is thrown. How about the following message?
```suggestion
query. When the breaker is triggered, an `org.elasticsearch.common.breaker.CircuitBreakingException`
is thrown and a descriptive error message is returned to the user.
```

Author: I'm torn between `an org......` and `a ...CircuitBreakingException` :D

Reviewer 2: Leaving out the package name or even Exception (which are developers facing implementation details) is preferred. Just an "error message" or similar is enough. Not all users are developers.

This <<circuit-breaker, circuit breaker>> can be configured using the following settings:

`breaker.eql_sequence.limit`::
(<<cluster-update-settings,Dynamic>>) The limit for circuit breaker used to restrict
the memory utilisation during the execution of an EQL sequence query. This value is
defined as a percentage of the JVM heap. Defaults to `50%`. If the
<<parent-circuit-breaker,parent circuit breaker>> is set to a value less than `50%`,
this setting uses that value as its default instead.

`breaker.eql_sequence.overhead`::

> (<<cluster-update-settings,Dynamic>>) A constant which all computed memory required
> for the execution of an EQL query is multiplied by to determine a final estimation.
> See <<circuit-breaker>>. Defaults to `1`.

Reviewer 3: More shortening.
```suggestion
(<<cluster-update-settings,Dynamic>>) A constant that sequence query memory
estimates are multiplied by to determine a final estimate. Defaults to `1`.
```

`breaker.eql_sequence.type`::
(<<static-cluster-setting,Static>>) The underlying type of the circuit breaker.
There are two valid options: `noop` and `memory`. `noop` means the circuit
breaker does nothing to prevent too much memory usage. `memory` means the

> circuit breaker tracks the memory used by trained models and can potentially

Reviewer: 'used by trained models' ?

Author: oh, thx! copy-paste error :(

> `breaker.eql_sequence.type`::
> (<<static-cluster-setting,Static>>) The underlying type of the circuit breaker.
> There are two valid options: `noop` and `memory`. `noop` means the circuit
> breaker does nothing to prevent too much memory usage. `memory` means the
> circuit breaker tracks the memory used by sequence queries and can potentially
> break and prevent `OutOfMemory` errors. The default value is `memory`.

Reviewer 3: It seemed odd to me that this wasn't a Boolean. It seems like you're disabling or enabling the breaker rather than definining its type.

```suggestion
`breaker.eql_sequence.type`::
(<<static-cluster-setting,Static>>) Circuit breaker type. Valid values are:

`memory` (Default):::
The breaker limits memory usage for EQL sequence queries.

`noop`:::
Disables the breaker.
```
